### PR TITLE
add task space for loss

### PR DESCRIPTION
**Pull request by chengaofei**
add task space for multi task loss

### CODE REVIEW

File: tzrec/models/rank_model.py [hunk 1]
```diff
@@ -216,24 +211,27 @@ def _loss_impl(
         else:
             raise ValueError(f"loss[{loss_type}] is not supported yet.")
 
-        if sample_weight_name:
-            losses[loss_name] = torch.mean(
-                losses[loss_name] * sample_weights
-            ) / torch.mean(sample_weights)
+        losses[loss_name] = torch.mean(losses[loss_name] * loss_weight)
```
Comment (tiankongdeguiji) [inline tzrec/models/rank_model.py]: - loss_weight should be None rather than ones tensor when do not use loss weight.
- reduction in _init_loss_impl should be none when do not use loss weight.

File: tzrec/models/multi_task_rank.py [hunk 2]
```diff
@@ -76,18 +76,28 @@ def loss(
...
             tower_name = task_tower_cfg.tower_name
             label_name = task_tower_cfg.label_name
-            sample_weight_name = (
-                task_tower_cfg.sample_weight_name
-                if task_tower_cfg.sample_weight_name
-                else ""
-            )
+            if task_tower_cfg.sample_weight_name:
+                loss_weight = batch.sample_weights[task_tower_cfg.sample_weight_name]
+                loss_weight = loss_weight / torch.mean(loss_weight)
```
Comment (tiankongdeguiji) [inline tzrec/models/multi_task_rank.py:115]: `normalize loss_weight` should move to line 93

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: tzrec/models/multi_task_rank.py [hunk 3]
```diff
@@ -76,18 +76,28 @@ def loss(
...
+            else:
+                loss_weight = torch.Tensor([1.0]).to(batch.labels[label_name].device)
+            loss_weight *= task_tower_cfg.weight
+            if task_tower_cfg.HasField("task_space_indicator_label"):
+                in_task_space = (
+                    batch.labels[task_tower_cfg.task_space_indicator_label] > 0
+                ).float()
+                loss_weight = loss_weight * (
```
Comment (tiankongdeguiji) [inline tzrec/models/multi_task_rank.py]: add a unit test for task space weight in multi_task_rank_test

File: tzrec/protos/tower.proto [hunk 4]
```diff
@@ -36,6 +36,12 @@ message TaskTower {
     optional float weight = 7 [default = 1.0];
     // sample weight for the task
     optional string sample_weight_name = 8;
+    // label name for indicating the sample space for the task tower
+    optional string task_space_indicator_label = 9;
```
Comment (yjjinjie) [inline tzrec/protos/tower.proto:40]: add unittest test for this new feature

File: tzrec/models/multi_task_rank.py [hunk 5]
```diff
@@ -76,18 +76,28 @@ def loss(
...
-                if task_tower_cfg.sample_weight_name
-                else ""
-            )
+            if task_tower_cfg.sample_weight_name:
+                loss_weight = batch.sample_weights[task_tower_cfg.sample_weight_name]
+                loss_weight = loss_weight / torch.mean(loss_weight)
+            else:
+                loss_weight = torch.Tensor([1.0]).to(batch.labels[label_name].device)
+            loss_weight *= task_tower_cfg.weight
+            if task_tower_cfg.HasField("task_space_indicator_label"):
```
Comment (tiankongdeguiji) [inline tzrec/models/multi_task_rank.py]: add doc for the param in MMoE, DBMTL, PLE models

File: tzrec/models/multi_task_rank.py (hunk 2 above)
Comment (chengaofei) [reply tzrec/models/multi_task_rank.py:115]: ok

File: tzrec/protos/tower.proto (hunk 4 above)
Comment (chengaofei) [reply tzrec/protos/tower.proto:40]: multi_task_rank_test.py has add
